sandbox/linux/controller.go

```diff
@@ -0,0 +1,318 @@
+package linux
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/api/runtime/task/v2"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	osinterface "github.com/containerd/containerd/pkg/os"
+	"github.com/containerd/containerd/pkg/process"
+	"github.com/containerd/containerd/pkg/stdio"
+	v2 "github.com/containerd/containerd/runtime/v2"
+	"github.com/containerd/containerd/runtime/v2/runc"
+	"github.com/containerd/containerd/sandbox"
+	"github.com/containerd/containerd/sys/reaper"
+	runcC "github.com/containerd/go-runc"
+	"github.com/containerd/typeurl"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+type sandboxController struct {
+	root          string
+	state         string
+	pauseLocation string
+	taskAddress   string
+	os            osinterface.OS
+	platform      stdio.Platform
+	sandboxes     map[string]runcSandbox
+}
+
+func NewSandboxController(ctx context.Context, root string, state string, pauseLocation string, taskAddress string) (sandbox.Controller, error) {
+	platform, err := runc.NewPlatform()
+	if err != nil {
+		return nil, err
+	}
+	signals := make(chan os.Signal, 32)
+	smp := []os.Signal{unix.SIGCHLD}
+	signal.Notify(signals, smp...)
+	go reap(ctx, signals)
+	if err := reaper.SetSubreaper(1); err != nil {
+		return nil, err
+	}
+	ec := reaper.Default.Subscribe()
+
+	controller := &sandboxController{
+		root:          root,
+		state:         state,
+		pauseLocation: pauseLocation,
+		taskAddress:   taskAddress,
+		platform:      platform,
+		sandboxes:     make(map[string]runcSandbox),
+		os:            osinterface.RealOS{},
+	}
+	if err := controller.recover(); err != nil {
+		return nil, err
+	}
+	go controller.processExits(ec)
+
+	return controller, nil
+}
+
+func (s *sandboxController) Start(ctx context.Context, sandbox *sandbox.Sandbox) (*sandbox.Sandbox, error) {
+	log.G(ctx).Debugf("start runc sandbox %s with a pause container", sandbox.ID)
+	sandbox.Spec.Process.Args = []string{"/pause"}
```

Comment:
The current implementation of runc's Sandboxer is still a task and shim process, right?

Reply:
there will be a pause container started by runc to provide the sandbox namespaces, but the task server is embeded in the containerd, so there is no shim process for runc pod.